content/gRPC/Go.md

```diff
@@ -0,0 +1,486 @@
+}
+{{</highlight>}}
+
+##### Generate the service
+
+```shell
+protoc -I rpc rpc/defs.proto --go_out=plugins=grpc:rpc
+```
+
+which should now generate the following directory structure
+
+```shell
+ls -R
+rpc
+
+./rpc:
+defs.pb.go	defs.proto
+```
+
+##### Service implementation
+
+In order to use the gRPC service, we need to implement the server.
+
+Create a file called `server.go` containing this code
+
+{{<highlight go>}}
+package main
+
+import (
+	"bytes"
+	"context"
+	"log"
+	"net"
+
+	"google.golang.org/grpc"
+
+	"oc.tutorials/ocgrpc/rpc"
+)
+
+type fetchIt int
+
+// Compile time assertion that fetchIt implements FetchServer.
+var _ rpc.FetchServer = (*fetchIt)(nil)
+
+func (fi *fetchIt) Capitalize(ctx context.Context, in *rpc.Payload) (*rpc.Payload, error) {
+	out := &rpc.Payload{
+		Data: bytes.ToUpper(in.Data),
+	}
+	return out, nil
+}
+
+func main() {
+	addr := ":9988"
+	ln, err := net.Listen("tcp", addr)
+	if err != nil {
+		log.Fatalf("gRPC server: failed to listen: %v", err)
+	}
+	srv := grpc.NewServer()
+	rpc.RegisterFetchServer(srv, new(fetchIt))
+	log.Printf("fetchIt gRPC server serving at %q", addr)
+	if err := srv.Serve(ln); err != nil {
+		log.Fatalf("gRPC server: error serving: %v", err)
+	}
+}
+{{</highlight>}}
+
+which you can run by
+
+```shell
+go run server.go
+```
+
+##### Client
+
+The client talks to the server via a gRPC channel, sending in bytes and getting back the output capitalized.
+
+The contents of `client.go` are as below:
+
+{{<highlight go>}}
+package main
+
+import (
+	"bufio"
+	"context"
+	"fmt"
+	"log"
+	"os"
+
+	"google.golang.org/grpc"
+
+	"oc.tutorials/ocgrpc/rpc"
+)
+
+func main() {
+	serverAddr := ":9988"
+	cc, err := grpc.Dial(serverAddr, grpc.WithInsecure())
+	if err != nil {
+		log.Fatalf("fetchIt gRPC client failed to dial to server: %v", err)
+	}
+	fc := rpc.NewFetchClient(cc)
+
+	fIn := bufio.NewReader(os.Stdin)
+	for {
+		fmt.Print("> ")
+		line, _, err := fIn.ReadLine()
+		if err != nil {
+			log.Printf("Failed to read a line in: %v", err)
+			return
+		}
+
+		ctx := context.Background()
+		out, err := fc.Capitalize(ctx, &rpc.Payload{Data: line})
+		if err != nil {
+			log.Printf("fetchIt gRPC client got error from server: %v", err)
+			continue
+		}
+		fmt.Printf("< %s\n\n", out.Data)
+	}
+}
+{{</highlight>}}
+
+and in another terminal, please run
+
+```shell
+go run client.go
+```
+and from typing you should be able to get back a response such as
+
+![](/images/ocgrpc-client.png)
+
+#### Instrumentation
+
+To gain insights to our service, we'll add trace and metrics instrumentation as follows
+
+##### Instrumenting the server
+
+We'll instrument the server by tracing as well as extracting gRPC metrics using the `ServerHandler`
+which will be registered as a grpc StatsHandler.
+
+{{<tabs Traces Metrics Combined>}}
+{{<highlight go>}}
+import "go.opencensus.io/trace"
+
+func (fi *fetchIt) Capitalize(ctx context.Context, in *rpc.Payload) (*rpc.Payload, error) {
```

Comment:
This code snippet is repeated in the block below, maybe just keep one copy of it?

Reply:
@songy23 we need that snippet in different steps as those are tabs. Please try running `hugo serve` and you'll see that those are tabs :)

Reply:
<img width="817" alt="screen shot 2018-08-07 at 10 59 55 pm" src="https://user-images.githubusercontent.com/4898263/43818903-a2158d50-9a95-11e8-8a1d-feab6d61e1ed.png">